overlord/snapstate/snapstate.go

```diff
@@ -81,6 +81,7 @@ func Install(s *state.State, snap, channel string, flags snappy.InstallFlags) (*
 // Note that the state must be locked by the caller.
 func Update(s *state.State, snap, channel string, flags snappy.InstallFlags) (*state.TaskSet, error) {
 	t := s.NewTask("update-snap", fmt.Sprintf(i18n.G("Updating %q"), snap))
+	t.Set("snap-setup-task", t.ID())
```

Comment:
Instead of adding self-referential metadata, just change TaskSnapSetup so it first looks for "snap-setup" instead.